File: backends/fastapi/app/services/capsule_service.py
"""
胶囊业务逻辑
封装创建、查询、列表、删除等核心操作
"""
import math
import secrets
import string
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy.orm import Session

from app.models import Capsule
from app.schemas import (
    CreateCapsuleRequest,
    CapsuleResponse,
    PageResponse,
)

CODE_CHARS = string.ascii_letters + string.digits  # A-Za-z0-9 (62 chars)
CODE_LENGTH = 8
MAX_RETRIES = 10


class CapsuleNotFoundException(Exception):
    """胶囊不存在"""
    def __init__(self, code: str):
        self.code = code
        super().__init__(f"胶囊不存在：{code}")


def _generate_code() -> str:
    """生成 8 位 base62 随机码"""
    return "".join(secrets.choice(CODE_CHARS) for _ in range(CODE_LENGTH))


def _generate_unique_code(db: Session) -> str:
    """生成唯一的胶囊码，最多重试 MAX_RETRIES 次"""
    for _ in range(MAX_RETRIES):
        code = _generate_code()
        if not db.query(Capsule).filter(Capsule.code == code).first():
            return code
    raise RuntimeError("无法生成唯一的胶囊码")


def _to_response_dict(
    capsule: Capsule,
    include_content: bool = False,
) -> dict:
    """将 Capsule 实体转换为响应字典，时间格式化为 ISO 8601 字符串"""
    now = datetime.now(timezone.utc)
    # 确保 open_at 有时区信息
    open_at = capsule.open_at
    if open_at.tzinfo is None:
        open_at = open_at.replace(tzinfo=timezone.utc)
    opened = now > open_at

    # 时间格式化为 ISO 8601 字符串
    open_at_str = open_at.isoformat().replace("+00:00", "Z")
    created_at_str = capsule.created_at.isoformat().replace("+00:00", "Z") if capsule.created_at else None

    result = {
        "code": capsule.code,
        "title": capsule.title,
        "creator": capsule.creator,
        "open_at": open_at_str,
        "created_at": created_at_str,
    }

    if include_content:
        result["content"] = capsule.content
    else:
        result["content"] = capsule.content if opened else None

    result["opened"] = opened
    return result


def create_capsule(db: Session, request: CreateCapsuleRequest) -> CapsuleResponse:
    """创建时间胶囊"""
    # openAt 必须在未来
    now = datetime.now(timezone.utc)
    if request.open_at <= now:
        raise ValueError("开启时间必须在未来")

    code = _generate_unique_code(db)

    capsule = Capsule(
        code=code,
        title=request.title,
        content=request.content,
        creator=request.creator,
        open_at=request.open_at,
        created_at=now,
    )
    db.add(capsule)
    db.commit()
    db.refresh(capsule)

    # 返回响应（不含 content）
    response_data = _to_response_dict(capsule, include_content=False)
    return CapsuleResponse(**response_data)


def get_capsule(db: Session, code: str) -> CapsuleResponse:
    """查询胶囊详情"""
    capsule = db.query(Capsule).filter(Capsule.code == code).first()
    if not capsule:
        raise CapsuleNotFoundException(code)
    response_data = _to_response_dict(capsule, include_content=False)
    return CapsuleResponse(**response_data)


def list_capsules(db: Session, page: int = 0, size: int = 20) -> PageResponse[CapsuleResponse]:
    """分页查询胶囊列表（管理员用）"""
    total = db.query(Capsule).count()
    total_pages = max(1, math.ceil(total / size))

    capsules = (
        db.query(Capsule)
        .order_by(Capsule.created_at.desc())
        .offset(page * size)
        .limit(size)
        .all()
    )

    content = [CapsuleResponse(**_to_response_dict(c, include_content=True)) for c in capsules]
    return PageResponse(
        content=content,
        total_elements=total,
        total_pages=total_pages,
        number=page,
        size=size,
    )


def delete_capsule(db: Session, code: str) -> None:
    """删除胶囊"""
    capsule = db.query(Capsule).filter(Capsule.code == code).first()
    if not capsule:
        raise CapsuleNotFoundException(code)
    db.delete(capsule)
    db.commit()

File: backends/gin/service/capsule_service.go
// Package service 胶囊业务逻辑
// 封装创建、查询、列表、删除等核心操作
package service

import (
	"crypto/rand"
	"errors"
	"fmt"
	"math"
	"math/big"
	"strings"
	"time"

	"github.com/alexhe/hellotime-gin/dto"
	"github.com/alexhe/hellotime-gin/model"
	"gorm.io/gorm"
)

const (
	codeChars  = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	codeLength = 8
	maxRetries = 10
)

var (
	ErrCapsuleNotFound = errors.New("胶囊不存在")
	ErrInvalidOpenAt   = errors.New("开启时间必须在未来")
	ErrCodeGeneration  = errors.New("无法生成唯一的胶囊码")
)

// generateCode 生成 8 位仅含大写字母和数字的随机码
func generateCode() (string, error) {
	var sb strings.Builder
	max := big.NewInt(int64(len(codeChars)))
	for i := 0; i < codeLength; i++ {
		n, err := rand.Int(rand.Reader, max)
		if err != nil {
			return "", err
		}
		sb.WriteByte(codeChars[n.Int64()])
	}
	return sb.String(), nil
}

// generateUniqueCode 生成唯一的胶囊码，最多重试 maxRetries 次
func generateUniqueCode(db *gorm.DB) (string, error) {
	for i := 0; i < maxRetries; i++ {
		code, err := generateCode()
		if err != nil {
			return "", err
		}
		var count int64
		db.Model(&model.Capsule{}).Where("code = ?", code).Count(&count)
		if count == 0 {
			return code, nil
		}
	}
	return "", ErrCodeGeneration
}

// formatTimeISO 格式化时间为 ISO 8601 字符串
func formatTimeISO(t time.Time) string {
	return t.UTC().Format("2006-01-02T15:04:05Z")
}

func parseStoredTime(value string) (time.Time, error) {
	normalized := strings.TrimSpace(strings.ReplaceAll(value, " ", "T"))
	if !strings.HasSuffix(normalized, "Z") && !strings.Contains(normalized[10:], "+") && !strings.Contains(normalized[10:], "-") {
		normalized += "Z"
	}
	return time.Parse(time.RFC3339Nano, normalized)
}

// toResponse 将 Capsule 实体转换为响应 DTO
func toResponse(capsule *model.Capsule, includeContent bool) dto.CapsuleResponse {
	now := time.Now().UTC()
	openAt, err := parseStoredTime(capsule.OpenAt)
	if err != nil {
		openAt = now
	}
	createdAt, err := parseStoredTime(capsule.CreatedAt)
	if err != nil {
		createdAt = now
	}
	openAt = openAt.UTC()
	createdAt = createdAt.UTC()
	opened := now.After(openAt)

	openAtStr := formatTimeISO(openAt)
	createdAtStr := formatTimeISO(createdAt)

	resp := dto.CapsuleResponse{
		Code:      capsule.Code,
		Title:     capsule.Title,
		Creator:   capsule.Creator,
		OpenAt:    openAtStr,
		CreatedAt: &createdAtStr,
		Opened:    &opened,
	}

	if includeContent {
		resp.Content = &capsule.Content
	} else if opened {
		resp.Content = &capsule.Content
	}
	// 未到时间且不强制包含 content 时，Content 为 nil

	return resp
}

// CreateCapsule 创建时间胶囊
func CreateCapsule(db *gorm.DB, req dto.CreateCapsuleRequest) (*dto.CapsuleResponse, error) {
	openAt, err := time.Parse(time.RFC3339, req.OpenAt)
	if err != nil {
		return nil, fmt.Errorf("openAt 格式错误，请使用 ISO 8601 格式: %w", err)
	}

	now := time.Now().UTC()
	if !openAt.After(now) {
		return nil, ErrInvalidOpenAt
	}

	code, err := generateUniqueCode(db)
	if err != nil {
		return nil, err
	}

	capsule := model.Capsule{
		Code:      code,
		Title:     req.Title,
		Content:   req.Content,
		Creator:   req.Creator,
		OpenAt:    formatTimeISO(openAt.UTC()),
		CreatedAt: formatTimeISO(now),
	}

	if err := db.Create(&capsule).Error; err != nil {
		return nil, err
	}

	resp := toResponse(&capsule, false)
	// 创建响应不含 opened 和 content
	resp.Opened = nil
	resp.Content = nil
	return &resp, nil
}

// GetCapsule 查询胶囊详情
func GetCapsule(db *gorm.DB, code string) (*dto.CapsuleResponse, error) {
	var capsule model.Capsule
	if err := db.Where("code = ?", code).First(&capsule).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, ErrCapsuleNotFound
		}
		return nil, err
	}

	resp := toResponse(&capsule, false)
	return &resp, nil
}

// ListCapsules 分页查询胶囊列表（管理员用）
func ListCapsules(db *gorm.DB, page, size int) (*dto.PageResponse, error) {
	var total int64
	db.Model(&model.Capsule{}).Count(&total)

	totalPages := int(math.Max(1, math.Ceil(float64(total)/float64(size))))

	var capsules []model.Capsule
	db.Order("created_at DESC").Offset(page * size).Limit(size).Find(&capsules)

	content := make([]dto.CapsuleResponse, len(capsules))
	for i, c := range capsules {
		content[i] = toResponse(&c, true)
	}

	return &dto.PageResponse{
		Content:       content,
		TotalElements: total,
		TotalPages:    totalPages,
		Number:        page,
		Size:          size,
	}, nil
}

// DeleteCapsule 删除胶囊
func DeleteCapsule(db *gorm.DB, code string) error {
	result := db.Where("code = ?", code).Delete(&model.Capsule{})
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return ErrCapsuleNotFound
	}
	return nil
}
